aten/src/ATen/native/cuda/block_reduce.cuh

```diff
@@ -27,7 +36,8 @@ __inline__ __device__ T BlockReduceSum(T val, T* shared) {
     shared[wid] = val;
   }
   __syncthreads();
-  val = (threadIdx.x < blockDim.x / C10_WARP_SIZE) ? shared[lid] : 0;
+  int n_warps = (blockDim.x + C10_WARP_SIZE - 1) / C10_WARP_SIZE;
```

Comment:
WARP_SHFL_DOWN above will correctly work only when all the threads in the warp participate https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions,
so blockDim.x should be a multiple of C10_WARP_SIZE (it's good to leave comment here about it), so this change shouldn't be necessary, and calling code should make sure that this condition is satisfied.

Reply:
Ah, you are right. That said, generalizing the function to work with functions not multiple of `C10_WARP_SIZE` shouldn't be too hard. For example, we could modify `WarpReduceSum` to something like:

```c++
for (int offset = (C10_WARP_SIZE >> 1); offset > 0; offset >>= 1) {
    T tmp = WARP_SHFL_DOWN(val, offset);
    // When (thread_id + offset >= block_size), tmp is undefined.
    val += (thread_id + offset < block_size) ? tmp : 0;
}
```

What do you think?

Reply:
I'd prefer to not do it, because it makes this code slower (due to additional checks) and less readable. The reduction is a performance critical part of many kernels, so the faster it is, the better. The "real" number of threads is multiple of warp size anyway (that's how cuda works), so it's better (more performant and more readable) to handle it at the kernel level with masked reads and writes for the threads that are outside of the working range.

Reply:
That does make sense. I shall do that.